### ISSUE TITLE
Not able to reproduce the results reported in the paper

### ISSUE THREAD

**Issue by kaiwinut** [matalvepu/HKT#6]
Hi, I'm trying to do some research on computational humor, and I found the [paper](https://ojs.aaai.org/index.php/AAAI/article/view/17534) you published in AAAI 2021 really inspiring 😀

I read & ran your code to see if I could fully understand the proposed method.
However, I've been struggling to reproduce the results you reported in the paper.
It would be really nice if you would please help my with it 🙇🏻

### Things I have tried so far

1. Clone the repository and run `python test_pretrained_models.py --dataset="humor" --max_seq_length=64 --cross_n_heads=2 --fusion_dim=300`
    - The result I got was identical to what you had in your README and your paper `Output: Accuracy: 77.36, F score: 77.36`

2. Next, I tried to train the model from scratch by running `python main.py --dataset "humor"`
    - This time I only got around 72.636% accuracy, I'll put the log file at the end

3. Then I figured maybe it was the hyper-parameters that were wrong, so I followed your README and tried `python main.py --dataset "humor" --max_seq_length 64 --cross_n_heads=2 --fusion_dim=300`
    - This time the accuracy was around 73.138%. Still, this is about 4% lower than the expected accuracy

4. I read your code and found some parts that I couldn't fully understand ( I'm new to this area, so please correct me if I'm wrong
    - In `models.py`, you implemented the `Attention` class without a fully connected layer at the end of the forward pass. I checked the paper and [Pytorch's implementation](https://github.com/pytorch/pytorch/blob/c6c207f620d01c02f192613c5a231f2581dc3997/torch/nn/functional.py#L5179) of Multi-Head Attention, and both had the fully connected layer. Was this intended?
    - Also in the `Attention` class (& in your paper), you [divided the square root of the `attention_head_size`](https://github.com/matalvepu/HKT/blob/main/models.py#L128) instead of `hidden_size` before apply the softmax function. Was this intended too?
    - Same in `models.py`, you [implemented](https://github.com/matalvepu/HKT/blob/main/models.py#L88) the `TransformerEncoder` with skip connections between each `TransformerLayer`. I tried to find an explanation to this in your paper, however as I read through the _"Acoustic, Visual and Humor Centric Feature Representations"_ section, I thought that you were using the original Transformer architecture proposed in the _"Attention is All You Need"_ paper.

5. Last, after reading your code, I decided to implement my own HKT class. I followed almost 99% of your code (except what I mentioned in 4.) to preprocess data and train the model. I also tried to train the unimodal models first, since you mentioned that during HKT's training, most of the gradients flow to the ALBERT model. However, I came across the following issues during training
    - I could not train any unimodal models except hcf_model. I used the hyper-parameters in `main.py` to train them. hcf_model eventually got around 54% accuracy after 100 epochs, which is still 2% lower than in the paper. Models of other modalities (vision & audio) keep predicting at 50% accuracy.

6. I examined the predictions that the model made, and found that some labels in the dataset are not quite intuitive. Here are two examples (I'm using UR-FUNNY v2)
    - Video clip 4841.mp4 should be a funny clip (at least to me), however it was labeled as unfunny
    - Video clip 15297.mp4 should be an unfunny clip (at least to me), however it was labeled as funny

### Also, could you provide
 - The actual code you executed to get the best parameters provided in this repository
 - The code you used to extract and process openface & covarep features from the raw_videos, since it was not provided in [UR-FUNNY's repository](https://github.com/ROC-HCI/UR-FUNNY) either
 - The version of transformers you are using in requirements.txt  <- Never mind I found out it was `3.1.0`

### Thank you in advance 😀

---
Output log of `python main.py --dataset "humor"`
```
Dataset Loaded:  humor
Model Loaded:  HKT
epoch:0,train_loss:0.6953265984393969, valid_loss:0.6421878102325624
Accuracy: 0.6428571428571429 F score: 0.62347153152674
epoch:1,train_loss:0.5947738088229123, valid_loss:0.5731234266873328
Accuracy: 0.7243460764587525 F score: 0.7240554543803989
epoch:2,train_loss:0.4977058132155603, valid_loss:0.5610591278922173
Accuracy: 0.7263581488933601 F score: 0.7259922345274968
epoch:3,train_loss:0.3723218550263834, valid_loss:0.6133877659036268
Accuracy: 0.6921529175050302 F score: 0.6878503462245042
epoch:4,train_loss:0.22379774349045103, valid_loss:0.683254836307418
Accuracy: 0.6891348088531187 F score: 0.6891187626066614
epoch:5,train_loss:0.11886008232127593, valid_loss:0.893123441646176
Accuracy: 0.6891348088531187 F score: 0.686583941700772
epoch:6,train_loss:0.06235623670773221, valid_loss:1.0024105654128137
Accuracy: 0.6800804828973843 F score: 0.6800804828973843
epoch:7,train_loss:0.03243555386644639, valid_loss:1.0433029950626436
Accuracy: 0.6871227364185111 F score: 0.6871179853487682
epoch:8,train_loss:0.023583831238666034, valid_loss:1.1845549681494314
Accuracy: 0.6810865191146881 F score: 0.6809732056506381
epoch:9,train_loss:0.012612289327447217, valid_loss:1.234051086969914
Accuracy: 0.6619718309859155 F score: 0.6619608830597264
epoch:10,train_loss:0.00853983307758174, valid_loss:1.307039903056237
Accuracy: 0.6690140845070423 F score: 0.6684388569861994
epoch:11,train_loss:0.005059894615419389, valid_loss:1.3626099276476569
Accuracy: 0.6659959758551308 F score: 0.6651994638738633
epoch:12,train_loss:0.004206045945771743, valid_loss:1.4162620978970681
Accuracy: 0.6609657947686117 F score: 0.6607505676476659
epoch:13,train_loss:0.0031856127021425333, valid_loss:1.465615730131826
Accuracy: 0.6589537223340041 F score: 0.6589692556453689
epoch:14,train_loss:0.002629277578353522, valid_loss:1.5126731652886636
Accuracy: 0.6529175050301811 F score: 0.651908113409665
epoch:15,train_loss:0.001913825499271473, valid_loss:1.568238744812627
Accuracy: 0.6589537223340041 F score: 0.6588035365891756
epoch:16,train_loss:0.0016636492695122323, valid_loss:1.599328944759984
Accuracy: 0.6559356136820925 F score: 0.6558520298743193
epoch:17,train_loss:0.001283554263726491, valid_loss:1.593209248156317
Accuracy: 0.6599597585513078 F score: 0.6596457036020429
epoch:18,train_loss:0.0011153057621260752, valid_loss:1.6151218923830217
Accuracy: 0.6579476861167002 F score: 0.6577814698431738
epoch:19,train_loss:0.0010259373919615726, valid_loss:1.6281702153144344
Accuracy: 0.6579476861167002 F score: 0.6577814698431738
main.py:270: UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor. (Triggered internally at  ../torch/csrc/utils/tensor_new.cpp:210.)
  all_visual = torch.tensor([f.visual for f in features], dtype=torch.float)
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:41<00:00,  4.69it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:04<00:00, 14.92it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:04<00:00, 14.28it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:40<00:00,  4.73it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:04<00:00, 14.68it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:04<00:00, 14.65it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:40<00:00,  4.74it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:04<00:00, 14.95it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:04<00:00, 14.56it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:44<00:00,  4.57it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:04<00:00, 14.82it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:04<00:00, 14.27it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:40<00:00,  4.76it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:04<00:00, 15.30it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:04<00:00, 14.37it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:42<00:00,  4.64it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:03<00:00, 16.97it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:04<00:00, 14.95it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:44<00:00,  4.56it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:03<00:00, 16.81it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:03<00:00, 16.90it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:42<00:00,  4.63it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:04<00:00, 14.94it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:04<00:00, 14.41it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:41<00:00,  4.69it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:04<00:00, 14.84it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:04<00:00, 14.52it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:40<00:00,  4.74it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:04<00:00, 15.02it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:04<00:00, 14.59it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:43<00:00,  4.61it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:04<00:00, 14.82it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:04<00:00, 14.55it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:40<00:00,  4.73it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:04<00:00, 15.05it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:04<00:00, 14.61it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:41<00:00,  4.71it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:04<00:00, 14.91it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:04<00:00, 14.15it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:41<00:00,  4.67it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:03<00:00, 17.54it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:03<00:00, 17.43it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:41<00:00,  4.67it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:04<00:00, 14.68it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:04<00:00, 14.72it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:41<00:00,  4.68it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:04<00:00, 15.01it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:04<00:00, 14.50it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:40<00:00,  4.76it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:04<00:00, 15.28it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:04<00:00, 14.59it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:41<00:00,  4.71it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:04<00:00, 14.36it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:04<00:00, 14.48it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:41<00:00,  4.70it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:04<00:00, 15.28it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:04<00:00, 14.60it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:41<00:00,  4.68it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:04<00:00, 14.92it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:04<00:00, 14.18it/s]
```

---
Output log of `python main.py --dataset "humor" --max_seq_length 64 --cross_n_heads=2 --fusion_dim=300`
```
Dataset Loaded:  humor
Model Loaded:  HKT
epoch:0,train_loss:0.6893243696384069, valid_loss:0.6689327315938088
Accuracy: 0.6156941649899397 F score: 0.5798496770922876
epoch:1,train_loss:0.5864339787043443, valid_loss:0.5560937716114905
Accuracy: 0.7313883299798792 F score: 0.7303534366764276
epoch:2,train_loss:0.49029286365423885, valid_loss:0.5705690477644244
Accuracy: 0.7233400402414487 F score: 0.722948624076859
epoch:3,train_loss:0.3603451849125764, valid_loss:0.6057111519959665
Accuracy: 0.716297786720322 F score: 0.715563573102748
epoch:4,train_loss:0.2141923082976782, valid_loss:0.7157446273392246
Accuracy: 0.7022132796780685 F score: 0.7021915790608104
epoch:5,train_loss:0.10871446408143565, valid_loss:0.8938600714168241
Accuracy: 0.7072434607645876 F score: 0.7071144598128328
epoch:6,train_loss:0.05261107459527386, valid_loss:0.9931491930638591
Accuracy: 0.6881287726358148 F score: 0.6878786571866311
epoch:7,train_loss:0.029843844935110136, valid_loss:1.1103526048602597
Accuracy: 0.6921529175050302 F score: 0.6910054283799271
epoch:8,train_loss:0.024705329059306218, valid_loss:1.1563182634812208
Accuracy: 0.6881287726358148 F score: 0.6880050142122577
epoch:9,train_loss:0.014177339874695157, valid_loss:1.2894619366815012
Accuracy: 0.6871227364185111 F score: 0.6866304871335809
epoch:10,train_loss:0.010691422071041805, valid_loss:1.3416767355895811
Accuracy: 0.6830985915492958 F score: 0.6828625559404394
main.py:270: UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor. (Triggered internally at  ../torch/csrc/utils/tensor_new.cpp:210.)
  all_visual = torch.tensor([f.visual for f in features], dtype=torch.float)
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:24<00:00,  5.63it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:03<00:00, 18.11it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:03<00:00, 18.05it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:24<00:00,  5.61it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:03<00:00, 18.51it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:03<00:00, 18.13it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:25<00:00,  5.57it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:03<00:00, 18.07it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:03<00:00, 17.91it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:25<00:00,  5.56it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:03<00:00, 17.77it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:03<00:00, 17.50it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:24<00:00,  5.60it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:03<00:00, 16.86it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:03<00:00, 16.08it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:25<00:00,  5.60it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:02<00:00, 21.14it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:03<00:00, 20.66it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:25<00:00,  5.57it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:03<00:00, 18.09it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:03<00:00, 18.18it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:23<00:00,  5.67it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:03<00:00, 20.08it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:03<00:00, 17.43it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:25<00:00,  5.57it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:03<00:00, 17.79it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:03<00:00, 17.69it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:24<00:00,  5.62it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:03<00:00, 15.96it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:03<00:00, 17.82it/s]
Iteration: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 476/476 [01:25<00:00,  5.57it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 62/62 [00:03<00:00, 18.59it/s]
Iteration: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 63/63 [00:03<00:00, 18.13it/s]
Iteration:  28%|████████████████████████████████▉                                                                                     | 133/476 [00:23<01:01,  5.57it/s]
Traceback (most recent call last):
  File "main.py", line 753, in <module>
    main()
  File "main.py", line 741, in main
    train(
  File "main.py", line 555, in train
    train_loss = train_epoch(
  File "main.py", line 384, in train_epoch
    model.zero_grad()
  File "/home/mil/wang-kai/.pyenv/versions/anaconda3-2020.11/envs/hkt/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1797, in zero_grad
    if p.grad.grad_fn is not None:
KeyboardInterrupt
```

**Comment by RE-N-Y** [comment]
Hi! I'm one of the co-authors of the paper. Perhaps, it's too late but I will try to address a couple of issues.

**1. Training not providing the same accuracy**
As is common in SOTA research, we ran a decent amount of hyperparameter searches of our method and the baselines to get the best performance. Again, there are a lot of moving parts such as random seed, learning rate, number of heads, hardware, etc. That's why we provide the exact weights to replicate the result. The same principle applies to unimodal models. Unless you run a decent amount of hyperparameter searches, you will not likely get the same level of accuracy.

**2. Architecture design**
- Dividing by sqrt of attention head size
Yes, that is perfectly standard in "Multi"head Attention models. Check [Karparthy's minGPT implementation](https://github.com/karpathy/minGPT/blob/master/mingpt/model.py) here for reference. Multihead attention is sort of a parallel version of attention where you run attention for each head and aggregate it in the end. 
So, if your feature size is 128, and you have 4 heads, you have 32 features for every 4 heads and you run each head in parallel. Thus, dividing by $\sqrt{32}$ not $\sqrt{128}$

- Skip connections
In general, skip connections are common techniques to improve the flow of gradient. It's just good to add in whenever you have a new layer (as long as it's not too exhaustive)

- MLP layers
Again, it's quite common in the literature to first process the data and apply MLP layers at the end. However, feel free to add extra MLP layers. Technically, "attention" usually strictly refers to the "attention" without the MLP part. I believe most people say it's a Transformer layer when an MLP component is also added.

**3. Funny videos are not funny**
Sadly, UR-FUNNY's labeling is not perfect. The dataset was made by processing the videos through a speech-to-text algorithm which also provides timestamps with laughter. Maybe the dataset could be improved with the latest algorithms like [whipser](https://github.com/openai/whisper). But because the video transcription is not perfect, there could be a few data points that wouldn't seem funny.

**4. Providing code for feature extraction**
It's been years since I wrote this paper so the code is probably buried in some Jupyter Notebook on our server.
Although, here are the link to [Openface](https://github.com/TadasBaltrusaitis/OpenFace) and [Covarep](https://covarep.github.io/covarep/)

Also, if your goal is to get a SOTA algorithm, there're alternative feature extraction algorithms to note a few:
1. [Mediapipe](https://google.github.io/mediapipe/)
2. [Any of audio embeddings from huggingface models](https://huggingface.co/docs/transformers/model_doc/audio-spectrogram-transformer)
3. [Data2Vec 2.0](https://ai.facebook.com/blog/ai-self-supervised-learning-data2vec/)

Hope that helps! Sorry that it took so long to respond, I don't get notifications from this repository.

**Comment by kaiwinut** [comment]
@RE-N-Y 
Hi, sorry for the late reply! I'm really grateful for your detailed response.

**1.** As you said, it might be the random seed that caused the difference in accuracy. 
**2.** Thanks for the explanations on the architecture! I've been reading some papers and I think I can now understand the concept behind it.
**3.** I also gathered some data eventually for the experiments in my research, and I understand that making the UR-FUNNY dataset is really hard work. Thank you for making these resource available!
**4.** Thanks for the links! I'm sure they'll come in useful in my future studies on multimodal humor.

Once again, thank you for the detailed reply and all the help. I learned a lot from your paper and the comments.
Wish you good luck in future projects!
